pandas/core/indexing.py

```diff
@@ -602,7 +602,7 @@ def _get_setitem_indexer(self, key):
         """
         Convert a potentially-label-based key into a positional indexer.
         """
-        if self.name == "loc":
+        if self.name == "loc" or com.is_bool_indexer(key):
```

Comment:
i haven't fully formed an opinion, but i _think_ another alternative would be to properly handle boolean masks in Block.setitem.

Reply:
You are right, loc is not failing, this means we do not need this or. Removed it.

Reply:
Lists can be handled in Block.setitem already, but arrays from dtype object don't work. Lists are killed by the validation mentioned above in ``pandas.core.indexers``.